An sxiv-like marks system

Flurrywinde opened the pull request:
I missed how sxiv has a marks system, so I put together one for pqiv. Currently, you can toggle marks (see them in montage mode), clear all marks, and pipe all marked files to an external `command()` by prefacing with `-` (e.g. in pqivrc, put a keybinding to `command(-cat)` (or other external command besides `cat`)).

Example keybinds to actually use it (add to pqivrc in both regular and montage mode sections):
```
m { toggle_mark() }
<Control>m { clear_marks() }
<Control>i { command(-cat) }
```

phillipberndt commented:
Cool! It'll take some time for me to have a proper look, sorry for that in advance. Will leave a few comments on code cleanups though.

phillipberndt commented on pqiv.c:
```diff
@@ -705,6 +705,8 @@ const struct pqiv_action_descriptor {
 	{ "move_window", PARAMETER_2SHORT },
 	{ "toggle_background_pattern", PARAMETER_INT },
 	{ "toggle_negate_mode", PARAMETER_INT },
+	{ "toggle_mark", PARAMETER_INT },
```
Why does this take a parameter? (Your code below looks like you intended to use this to mark specific images - what's the use case for that?)

Flurrywinde replied:
Just leaving room for future development, I guess, since maybe someone might have a use for toggling an arbitrary mark. Should I remove it? (I can't think of a use case at the moment.)

phillipberndt commented on pqiv.c:
```diff
@@ -780,6 +782,12 @@ void draw_current_image_to_context(cairo_t *cr);
+void clear_marks();
+char *get_all_marked();
+#if defined (__GLIBC__)
+	FILE *open_memstream(char **ptr, size_t *sizeloc);
```
Use https://developer.gnome.org/glib/stable/glib-Strings.html for compatibility

phillipberndt commented on pqiv.c:
```diff
@@ -1521,6 +1529,7 @@ void load_images_handle_parameter(char *param, load_images_state_t state, gint d
 	// Check for memory image
 	if(state == PARAMETER && g_strcmp0(param, "-") == 0) {
 		file = g_slice_new0(file_t);
+		file->marked = -1;
```
If you'd make the field a boolean you wouldn't have to initialize to -1

phillipberndt commented on pqiv.c:
```diff
@@ -3530,6 +3540,23 @@ void apply_external_image_filter(gchar *external_filter) {/*{{{*/
 	}
+	else if(external_filter[0] == '<') {
+		char *marklist = get_all_marked();
+		// argv[2] = apply_external_image_filter_prepare_command(external_filter + 1);
```
Please remove the commented lines

phillipberndt commented on pqiv.c:
```diff
@@ -3530,6 +3540,23 @@ void apply_external_image_filter(gchar *external_filter) {/*{{{*/
+		// printf("%s\n", argv[2]);
+		// This does it a different way, so rest of argv not used.
+		// if(g_spawn_async(NULL, argv, NULL, 0, NULL, NULL, NULL, &error_pointer) == FALSE) {  // For diff between this and popen, see: https://stackoverflow.com/questions/219138/get-command-output-in-pipe-c-for-linux
+		FILE *fp = popen(external_filter + 1, "w");
```
Please use the g_spawn_async_with_pipes function I used to implement `|` instead.

phillipberndt commented on pqiv.c:
```diff
@@ -8095,4 +8147,51 @@ int main(int argc, char *argv[]) {
+	// Not using id for now. Just mark current.
+
+	if(application_mode == DEFAULT) {
+		if(FILE(current_file_node)->marked == 1) {
```
After switching to boolean you can remove the if and just negate the value

phillipberndt commented on pqiv.c:
```diff
@@ -8095,4 +8147,51 @@ int main(int argc, char *argv[]) {
+char *get_all_marked() {
+	char* result = NULL;
+	size_t resultSize = 0;
+	FILE* stream = open_memstream(&result, &resultSize);
```
See above, please use glib strings instead

Flurrywinde commented:
Great! I'll try to add these changes soon.

Flurrywinde commented:
Ok, I have implemented the requested changes. I also changed the prefix character in `command()` from `<` to `-`. (Just in case it might someday conflict with pango markup.) I should be able to add more features fairly soon, like reversing marks, goto next and previous marked file, and maybe also an option to make the slideshow show only marked files. (Should these be a new PR or added to this one? Sorry, I'm new to git and github.)

I was a little confused about `waitpid()`'s status. (See comment in the code.) I hope I got everything right.

phillipberndt commented on pqiv.c:
```diff
@@ -3530,6 +3535,50 @@ void apply_external_image_filter(gchar *external_filter) {/*{{{*/
+				g_printerr("Failed writing to stdin of %s\n", argv[2]);
+			}
+			close(child_stdin);
+			gint status = 0;  // When left uninitialized, external command reports exiting due to signal 95 (exit statuses have been 233, 201, 217). Why?
```
This is because waitpid fails with `errno == 10 (ECHILD)`, which is because `G_SPAWN_DO_NOT_REAP_CHILD` is not set, so glib will automatically wait for the child.

Interesting. This behaviour seems to have changed. The `G_SPAWN_DO_NOT_REAP_CHILD` below is not necessary either then (or the other way around: the waitpid isn't.

phillipberndt commented on pqiv.c:
```diff
@@ -4767,6 +4816,18 @@ gboolean window_draw_thumbnail_montage(cairo_t *cr_arg) {/*{{{*/
 			}
 
+			// Marks
+			if(thumb_file->marked) {
```
Those marks are really hard to see for some images. How about changing the [blend mode](https://cairographics.org/operators/) to take the underlying image into account, and then do something like negating the image?

phillipberndt commented on pqiv.c:
```diff
@@ -7987,6 +8060,40 @@ gboolean inner_main(void *user_data) {/*{{{*/
+}/*}}}*/
+void toggle_mark() {/*{{{*/
+	if(application_mode == DEFAULT) {
+		FILE(current_file_node)->marked = !FILE(current_file_node)->marked;
```
This should update the status message to give the user feedback that the mark was toggled.

(Also, is there some meaningful way to always tell the user in default mode as well that an image is marked? Maybe prefix or suffix the status message with some symbol?)

phillipberndt commented:
FYI, I did those changes myself and finally merged the branch. Thanks a lot for your contribution!

Flurrywinde commented:
Awesome! I worked on it a bit but was having trouble figuring out cairo, and then, you know, real life. I'll check out the release. Thanks for this great program. :-)
